memory-db/src/lib.rs

```diff
@@ -160,19 +160,151 @@ impl<H, KF, T> Eq for MemoryDB<H, KF, T>
 pub trait KeyFunction<H: KeyHasher> {
 	type Key: Send + Sync + Clone + hash::Hash + Eq;
 
-	fn key(hash: &H::Out, prefix: &[u8]) -> Self::Key;
+	fn key(hash: &H::Out, prefix: Prefix) -> Self::Key;
+}
+
+
+/// Make database key from hash and prefix.
+///
+/// Prefix could be ordered, but key hash appending will break this property.
+/// If true ordering is needed we would need a fix length encoding for prefix
+/// (so a maximum length and additional byte (or chunk of prefix encoded)).
+pub fn prefixed_key<H: KeyHasher>(key: &H::Out, prefix: Prefix) -> Vec<u8> {
+	let mut prefixed_key = Vec::with_capacity(key.as_ref().len() + prefix.0.len() + 1);
+	prefixed_key.extend_from_slice(prefix.0);
+	if (prefix.1).0 > 0 {
+		prefixed_key.push((prefix.1).1);
+		prefixed_key.push((prefix.1).0); // put size to avoid any possible collision
+	}
+	prefixed_key.extend_from_slice(key.as_ref());
+	prefixed_key
+}
+
+#[test]
+fn test_ordered_prefixed_key() {
+	use keccak_hasher::KeccakHasher;
+	#[derive(Debug, Clone, Default)]
+	struct Opt;
+	impl IntoIterator for Opt {
+		type Item = usize;
+		type IntoIter = std::iter::Repeat<usize>;
+		fn into_iter(self) -> Self::IntoIter {
+			std::iter::repeat(1)
+		}
+	}
+	impl OrderedPrefixedKeyChunk for Opt { }
+	let prefixes = vec![
+		(vec![],(0u8, 0x00u8), [1u8;32]),
+		(vec![],(1u8, 0x00u8), [1u8;32]),
+		(vec![0u8],(0,0), [0;32]),
+		(vec![],(1u8, 0x10u8), [1u8;32]),
+		(vec![0x10u8],(0,0), [0;32]),
+		(vec![0x11u8],(0,0), [0;32]),
+	];
+	let oc: std::collections::BTreeSet<_> = prefixes.iter().map(|(p1, p2, k)| {
+		ordered_prefixed_key::<KeccakHasher, Opt>(&k.clone().into(), (&p1[..], p2.clone()))
+	}).collect();
+
+	for ((p1, p2, k),b) in prefixes.iter().zip(oc.iter()) {
+		assert_eq!(
+			&ordered_prefixed_key::<KeccakHasher, Opt>(&k.clone().into(), (&p1[..], p2.clone())),
+			b,
+		);
+	}
+}
+/// Parameterize the size of successive chunks when using
+/// `ordered_prefixed_key`.
+/// Length of each chunk is minimum 1, maximum 254.
+pub trait OrderedPrefixedKeyChunk: IntoIterator<Item = usize> + Clone + MaybeDebug + Default {
+	/// Estimate a final length given a input length.
+	fn estimate_enc_len(input: usize) -> usize { input }
 }
 
 /// Make database key from hash and prefix.
-pub fn prefixed_key<H: KeyHasher>(key: &H::Out, prefix: &[u8]) -> Vec<u8> {
-	let mut prefixed_key = Vec::with_capacity(key.as_ref().len() + prefix.len());
-	prefixed_key.extend_from_slice(prefix);
+///
+/// The scheme allows byte ordering of key depending on prefix (can iterate
+/// trie with it).
+/// Every chunk we add a 2 byte (could be reduce to one if storing directly a
+/// total number of chunks), the number of written bytes (last prefix padded one included), and
+/// the number of bytes in the prefix padding byte (note that for aligned prefix we add a
+/// theorically useless 0 to keep a simple trait).
+pub fn ordered_prefixed_key<H, C>(key: &H::Out, prefix: Prefix) -> Vec<u8> 
```

Comment:
Do we really need this? Why is iteration not possible with current prefix scheme? It should be

Reply:
We do not really need this (this implementation is linked nowhere and I can consider removing it).

Current prefix does not allow direct kvdb iteration because we cannot say if we are comparing prefix or node hash, so we need to encode length of the prefix somehow, but encoding it tends to break the simple byte array comparison thus this complicated scheme.
But the main point is that being able to iterate on the trie nodes from kvdb have many limitation that makes it not interesting in most case:
- for state it iterate over all state nodes (we can see the history of value with a same prefix but helas it is not the same as value with same key), so for a collection containing multiple state of a trie it is not really usefull.
- for trie that got a single state in a single collection (or a with a common prefix), the iteration should work.

Reply:
I'd remove it. There's no use case for querying node by hash or iterating all nodes I believe. All we need is tree iteration, that is walking the tree starting from a root node down, not just enumerating all nodes. And that should work with unordered prefix.